### PR TITLE
Add date parsing to S3 info and infoList functions

### PR DESCRIPTION
**Pull request by dwsteele**
This will be broken up into a few commits, but I wanted to show the context of how the code will be used.

The main thing is that `common/time.c` has been updated to include functions to convert data/time parts into epoch time without using and of the standard C functions.  This is to avoid the well-known oddities of those functions plus concerns about behavioral differences on various platforms.

We should probably encapsulate this logic in a DateTime object at some point, but that will require more extensive changes.

So, the most important changes to look at are in `common/time.c`, but have a look at the rest as well.

### CODE REVIEW

File: test/src/module/common/timeTest.c [hunk 1]
```diff
@@ -31,5 +32,51 @@ testRun(void)
...
+        TEST_ERROR(timePartsValid( 0,  0, 60), FormatError, "invalid time 00:00:60");
+        TEST_RESULT_VOID(timePartsValid(23, 59, 59), "valid time");
+
+        // -------------------------------------------------------------------------------------------------------------------------
+        TEST_TITLE("time from parts");
+
+        TEST_RESULT_INT(epochFromParts(1970, 1, 1, 0, 0, 0), 0, "beginning of epoch");
+        TEST_RESULT_INT(epochFromParts(1996, 3, 3, 3, 3, 3), 825822183, "march of leap year");
+        TEST_RESULT_INT(epochFromParts(2000, 4, 4, 4, 4, 4), 954821044, "april of non-leap year");
+        TEST_RESULT_INT(epochFromParts(2038, 1, 19, 3, 14, 7), INT_MAX, "end of 32-bit signed int epoch");
```
Comment (cmwshang) [inline test/src/module/common/timeTest.c]: So do we need to somehow handle or throw an error for 32-bit systems after the Jan 19, 2038 date?

File: src/storage/s3/storage.c [hunk 2]
```diff
@@ -586,6 +587,21 @@ typedef struct StorageS3InfoListData
     void *callbackData;                                             // User-supplied callback data
 } StorageS3InfoListData;
 
+// Helper to convert S3 not-quite-ISO-8601 times to time_t
```
Comment (cmwshang) [inline src/storage/s3/storage.c]: I think we should define what "not-quite-ISO-8601" actually means - it seems to me it only accepts YYYY-MM-DDTHH:MM:SSZ format (optional .msec after SS).

File: src/common/io/http/common.c [hunk 3]
```diff
@@ -3,8 +3,49 @@ Http Common
...
 
+#include <string.h>
+
 #include "common/debug.h"
 #include "common/io/http/common.h"
+#include "common/time.h"
+
+/**********************************************************************************************************************************/
+time_t
+httpLastModifiedToTime(const String *lastModified)
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/common/io/http/common.c]: We currently only support HTTP 1.1 so I'm not worried about older format.  Any S3 TLS server must be HTTP 1.1.

I'll add the RFC section that we are coding to here.

File: src/common/io/http/common.c [hunk 4]
```diff
@@ -3,8 +3,49 @@ Http Common
...
+    FUNCTION_TEST_END();
+
+    time_t result = 0;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Find the month
+        static const char *monthList[] = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
+
+        const char *month = strPtr(strSubN(lastModified, 8, 3));
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/common/io/http/common.c]: No, RFC7231 requires a leading 0 in this case.

File: src/common/io/http/common.c [hunk 5]
```diff
@@ -3,8 +3,49 @@ Http Common
...
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(STRING, lastModified);
+    FUNCTION_TEST_END();
+
+    time_t result = 0;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Find the month
+        static const char *monthList[] = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/common/io/http/common.c]: Yes, RFC7231 specifies that all months abbr are in English.

File: test/src/module/common/timeTest.c [hunk 6]
```diff
@@ -31,5 +32,51 @@ testRun(void)
...
+        // -------------------------------------------------------------------------------------------------------------------------
+        TEST_TITLE("day of year");
+
+        TEST_RESULT_INT(dayOfYear(1995, 3, 1), 60, "1995-03-01 day of year");
+        TEST_RESULT_INT(dayOfYear(1996, 3, 1), 61, "1996-03-01 day of year");
+
+        // -------------------------------------------------------------------------------------------------------------------------
+        TEST_TITLE("date parts valid");
+
+        TEST_ERROR(datePartsValid(1969,  0,  0), FormatError, "invalid date 1969-00-00");
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply test/src/module/common/timeTest.c]: Test updated as requested.

File: test/src/module/common/timeTest.c (hunk 1 above)
Comment (dwsteele) [reply test/src/module/common/timeTest.c]: We'll worry about that in 18 years.  Probably we'll just disallow builds on systems with 32-bit time_t.

Note that this is not a backrest issue  -- everything will work with a properly sized time_t.

File: src/storage/s3/storage.c (hunk 2 above)
Comment (dwsteele) [reply src/storage/s3/storage.c]: Updated comment.

**Comment by dwsteele** [comment]
Committed at a08298ce1b, d2fb4f977, and 61538f932c5.  Thank you for the review.

**Pull request closed without merging**